man/man3/GPTLprocess_namelist.3
.TH gptlprocess_namelist 3 "May, 2020" "GPTL"

.SH NAME
gptlprocess_namelist \- Parse an input Fortran namelist and call
.B gptlsetoption() 
and/or 
.B gptlsetutr() 
appropriately.

.SH SYNOPSIS
.B Fortran Interface:
.nf
use gptl
subroutine gptlprocess_namelist (character(len=*) filename, integer unitno, integer ret)
.fi

.SH DESCRIPTION
.B gptlprocess_namelist()
is a utility subroutine provided for convenience to Fortran applications. It
allows the specification of input to
.B gptlsetoption()
and/or
.B gptlsetutr()
via namelist input. This avoids the necessity of recompiling and relinking
the user application when changing
.B GPTL
options. The routine is a subroutine rather than a function to
get around problems encountered when gptl.inc declares the
function to be external.

.SH ARGUMENTS
.TP
.I filename
-- file name containing the namelist (namelist group &gptlnl).

.TP
.I unitno
-- Fortran unit number to use when opening the namelist file.

.TP
.I ret
-- output return code

.SH RESTRICTIONS
.B gptlinitialize()
must 
.B NOT
have been called. Generally the user invokes
.B gptlinitialize()
right after calling
.B gptlprocess_namelist().

.SH RETURN VALUES
On success, 0 is returned in output argument
.B ret.
On error, 
.B ret
is negative and a descriptive message
printed. As long as the namelist file can be opened and read, 
.B ret 
is set to 0. This is true even if some of the individual settings fail (e.g. if
a requested PAPI event is not available).

.SH EXAMPLES
An example namelist (group name &gptlnl) is shown below. The same file is available in
ftests/gptnl. When 
.B gptlprocess_namelist()
is called, it parses the namelist and calls the
appropriate GPTL functions accordingly. Namelist variable names match
their counterparts in calls to
.B gptlsetoption() 
and 
.B gptlsetutr(), 
with the leading "gptl" left off. For example, the namelist can be used to
enable printing of overhead statistics by setting overhead = .true.; the same
thing is accomplished with the GPTL API by invoking 
.B gptlsetoption (gptloverhead, 1).
.nf         

&gptlnl
! These settings are all the opposite of the default
 sync_mpi        = .true.
 wall            = .false.
 cpu             = .true.
 abort_on_error  = .true.
 overhead        = .true.
 depthlimit      = 1
 verbose         = .true.
 percent         = .true.
 persec          = .false.
 multiplex       = .true.
 dopr_preamble   = .false.
 dopr_threadsort = .false.
 dopr_multparent = .false.
 dopr_collision  = .false.

! These use character variables instead of integer to avoid 
! "magic number" settings in the namelist

 utr             = 'nanotime'
 print_method    = 'full_tree'      ! default
!print_method    = 'first_parent'
!print_method    = 'last_parent'
!print_method    = 'most_frequent'
 eventlist       = 'PAPI_FP_OPS','GPTL_CI' ! enable 2 events
/

.fi

.SH SEE ALSO
.BR GPTLsetoption "(3)" 
.BR GPTLsetutr "(3)" 


ChangeLog
gptl3_6_2: print_memusage converts to MB by default (if possible).
	   Added print_memusage test to testbasics.F90
	   Added macros.make.bluegene (tested at ORNL)
	   Bugfix for process_namelist.F90: Needs to close opened file
	   on error.
	   Bugfix for gptl_papilibraryinit (Fortran): returns an int.
	   Changed LINUX ifdef to HAVE_SLASHPROC. Not all Linux systems
	   have /proc/pid/statm.
	   
gptl3_6_1: Bugfix for auto-profiling MPI_Recv: it could hang in gptl3_6.
	   Auto-profiling support for additional MPI routines.
	   Better estimates of bytes transfer for some auto-profiled
	   MPI routines.
gptl3_6:   Support for auto-profiling some MPI routines.
gptl3_5_3: Makefile simplification. Can now run "make" from ctests/ and
	   ftests/. Initial set of PMPI wrappers, and synchronize option.
gptl3_5_2: Bugfix for when omp_get_max_threads() returns zero.
gptl3_5_1: No more OMP pragmas: PAPI events started from get_thread_num().
	   Thus GPTL built with PTHREADS=yes now works when called from
	   OpenMP applications.
gptl3_4_7: Option to not run summary at all.
gptl3_4_6: totent counts collisions not entries.
	   Add target "testnompi" to only run tests that don't require MPI.
gptl3_4_5: Add realloc memory checks. Fix typo in uninstall target. 
	   Bugfix for instrumented codes when orphan encounters
	   parent. Print additional hash table stats. Decrease MAX_AUX
	   from 16 to 8 to save memory.
gptl3_4_4: Added GPTLevent_name_to_code and GPTLevent_code_to_name. Add
	some derived events. More description to print.
gptl3_4_2: Enable "make test". Error call to stop_instr won't invoke null
        pointer to function.
gptl3_4_1: Fix tests for robustness. Move initialized check in GPTLstop
	higher to avoid one of the error tests in ftests/errtest
gptl3_4: Put back GPTLpr_summary. Some people were using it.
	 Get ctests/ and ftests/ into a usable state.
	 Proper handling of input defines by Fortran.
gptl3_2: Handle multiple parents
	 Add parsegptlout.pl for handling MPI output
	 Add capability for derived counters (currently just IPC and CI)
	 Add GPTL_PAPIlibraryinit to allow enabling PAPI native events
	 Junk autoconf in favor of macros.make ("suggestions" retained as
           an autoconf-generated script to suggest settings)
	 Vastly faster start/stop for auto-instrumented codes
	 hex2name.pl properly indents output
	 Efficiency enhancements (e.g. no character copying on start/stop)
	 Use stack to determine parent, for simplicity, efficiency, correctness
	 Arrange timers in tree for proper handling of parent/child
gptl2_16: Rename GPTLpr_mpistats to GPTLpr_summary. Make it behave
	  reasonably even when MPI not enabled.
	  Close print files after writing to them.
gptl2_14: Add GPTLpr_mpistats entry point. 
	  Check string starts with "PAPI_" before lopping off.
	  Remove "inline" attribute to timing functions--they're probably
	  not inlined anyway.
gptl2_13: verbose = false by default. Wrap informational printout in 
	  test on "verbose". Add verbose option to gptl_papi.c
	  Modified behavior of scale.c
gptl2_11: Man pages. Default narrowprint to true. Fix minor formatting
	  bug in GPTLpr.
gptl2_10: Add GPTLpercent (of 1st timers[0]), and GPTLpersec (for PAPI).
	  Add timingModule.F90
gptl2_9: Update camgptl.inc to match gptl.inc (forgot in gptl2_8)
gptl2_8: Add option to retain parent/child relationships for printing.
	 Change behavior of ambiguous timers to put a "*" in column 1.
gptl2_7: Add GPTLget_nregions and GPTLget_regionname
gptl2_6: Add GPTLquerycounters (C and Fortran)
gptl2_5: Enable native events
gptl2_3: Minor changes to papiomptest.F
gptl2_2: Multiplexing: can enable, disable, or always enable.
	 GPTLnarrowprint (fewer digits)
	 Add cmd-line PAPI option arg to utrtest.
	 papiomptest no longer does malloc: caused trouble on SGI
gptl1_7: Split tests/ into ctests/ and ftests.
	 Fix logic on Fortran underscoring.
	 Return error when selected PAPI option not available.
	 Further fix (1.6 was not complete) on PAPI overhead.
gptl1_6: - Fix (bogus) print occurring from PAPI when overheadstats disabled.
	 - Add get_memusage and memusage test.
	 - Add "basic" test.
	 - Hack aclocal.m4 for openMP flags.
gptl1_5: - Fixed "install" target.  
	 - Add NANOTIME (Linux only) option.
	 - Replace gettimeofday with generic utr (underlying timing routine)
	 option.  
	 - Remove DIAG ifdef option.  
	 - Interpret nested "start" calls as recursion.  
	 - In NUMERIC mode, shift tag by 2 bits.
gptl1_4: - Added NUMERIC_TIMERS (configure --enable-numeric) for speedup.
	 If enabled, user args to start/stop MUST be numeric tags rather
	 than character strings.  Intended primarily to be used with
	 automatic code instrumentation (g++ -finstrument-functions), 
	 and tests/jr-resolve.pl to back out a name.
	 - Added "numeric" to tests/ in order to test --enable-numeric.
	 - Remove HASH as an option (it is now assumed) for readability.
	 - Moved some OMP code from threadutil.c to gptl.c strictly for
	 efficiency (inlining).  Hated to do it because the code really
	 doesn't belong there.
	 - Added jr-resolve.pl to tests/ to back out symbol name from
	 address. 
gpt1_2: Changed GPT to GPTL to avoid overlap with other utilities named GPT...
gpt1_1: Added README and man pages.  Better commenting.  Fixed bug in min
        calcs in GPTstop.
gpt1_0: Added PAPI support.  For whatever reason, on IBM get negative
	increments occasionally.
gpt0_3: Added overflow diagnostic novfl
	Migrate threading ops to threadutil.c
	Remove useless calcs from GPTstop
	Added iteration option to depth test
	overhead test is now a C code and tests potentially expensive
          calcs embedded in the library
	Removed exp, expensive, timetimeing tests
	Reenabled pthreads
	Fix SGI complaints about arg lists
	Changed depth array to mitigate false sharing
	Added DISABLE_TIMERS ifdef
	autoconf bugfix
	Change microsecond accumulator test from 1000000 to 10000000
gpt0_2: Remove unused components in Timer. Make "name" component a fixed

gptlprocess_namelist.f90
subroutine gptlprocess_namelist (filename, unitno, outret)
!
! $Id: gptlprocess_namelist.f90,v 1.6 2009-01-04 21:14:41 rosinski Exp $
!
! Author: Jim Rosinski
!
! Utility subroutine processes namelist group &gptlnl and makes appropriate 
! calls to gptlsetoption() and/or gptlsetutr().
!
! To follow GPTL conventions this should be a function not a subroutine.
! But 'include ./gptl.inc' and then setting function gptlprocess_namelist
! to a return value causes compiler to barf because the function is declared 
! 'external' in the header. So set return value in output arg 'outret' instead.
!
  implicit none

  character(len=*), intent(in) :: filename  ! Input file containing &gptlnl
  integer, intent(in) :: unitno             ! Fortran unit number to open
  integer, intent(out) :: outret            ! Output return code

  include './gptl.inc'

  integer :: j    ! loop index
  integer :: ios  ! status return from file open
  integer :: code ! event code
  integer :: ret  ! return value
  integer, parameter :: maxevents = 99 ! space to hold more than enough events

! Default values for namelist variables

  logical, parameter :: def_wall            = .true.
  logical, parameter :: def_cpu             = .false.
  logical, parameter :: def_abort_on_error  = .false.
  logical, parameter :: def_overhead        = .true.
  integer, parameter :: def_depthlimit      = 99999    ! Effectively unlimited
  logical, parameter :: def_verbose         = .false.
  logical, parameter :: def_narrowprint     = .true.
  logical, parameter :: def_percent         = .false.
  logical, parameter :: def_persec          = .true.
  logical, parameter :: def_multiplex       = .false.
  logical, parameter :: def_dopr_preamble   = .true.
  logical, parameter :: def_dopr_threadsort = .true.
  logical, parameter :: def_dopr_multparent = .true.
  logical, parameter :: def_dopr_collision  = .true.
  character(len=16), parameter :: def_print_method = 'most_frequent   '
  character(len=16), parameter :: def_utr          = 'gettimeofday    '

! Namelist values: initialize to defaults

  logical :: wall            = def_wall
  logical :: cpu             = def_cpu
  logical :: abort_on_error  = def_abort_on_error
  logical :: overhead        = def_overhead
  integer :: depthlimit      = def_depthlimit
  logical :: verbose         = def_verbose
  logical :: narrowprint     = def_narrowprint
  logical :: percent         = def_percent
  logical :: persec          = def_persec
  logical :: multiplex       = def_multiplex
  logical :: dopr_preamble   = def_dopr_preamble
  logical :: dopr_threadsort = def_dopr_threadsort
  logical :: dopr_multparent = def_dopr_multparent
  logical :: dopr_collision  = def_dopr_collision
  character(len=16) :: print_method    = def_print_method
  character(len=16) :: utr             = def_utr
  character(len=16) :: eventlist(maxevents) = (/('                ',j=1,maxevents)/)
  
  namelist /gptlnl/ wall, cpu, abort_on_error, overhead, depthlimit, &
                    verbose, narrowprint, percent, persec, multiplex, &
                    dopr_preamble, dopr_threadsort, dopr_multparent, dopr_collision, &
                    print_method, eventlist, utr

  open (unit=unitno, file=filename, status='old', iostat=ios)
  if (ios /= 0) then
    write(6,*)'gptlprocess_namelist: cannot open namelist file ', filename
    outret = -1
    return
  end if

  read (unitno, gptlnl, iostat=ios)
  if (ios /= 0) then
    write(6,*)'gptlprocess_namelist: failure reading namelist'
    outret = -1
    return
  end if

! Set options for user-defined values which are not default.
! Do verbose and abort_on_error first

  if (verbose .neqv. def_verbose) then
    if (verbose) then
      ret = gptlsetoption (gptlverbose, 1)
    else
      ret = gptlsetoption (gptlverbose, 0)
    end if
  end if

  if (abort_on_error .neqv. def_abort_on_error) then
    if (abort_on_error) then
      ret = gptlsetoption (gptlabort_on_error, 1)
    else
      ret = gptlsetoption (gptlabort_on_error, 0)
    end if
  end if

  if (wall .neqv. def_wall) then
    if (wall) then
      ret = gptlsetoption (gptlwall, 1)
    else
      ret = gptlsetoption (gptlwall, 0)
    end if
  end if

  if (cpu .neqv. def_cpu) then
    if (cpu) then
      ret = gptlsetoption (gptlcpu, 1)
    else
      ret = gptlsetoption (gptlcpu, 0)
    end if
  end if

  if (overhead .neqv. def_overhead) then
    if (overhead) then
      ret = gptlsetoption (gptloverhead, 1)
    else
      ret = gptlsetoption (gptloverhead, 0)
    end if
  end if

  if (depthlimit /= def_depthlimit) then
    ret = gptlsetoption (gptldepthlimit, depthlimit)
  end if

  if (narrowprint .neqv. def_narrowprint) then
    if (narrowprint) then
      ret = gptlsetoption (gptlnarrowprint, 1)
    else
      ret = gptlsetoption (gptlnarrowprint, 0)
    end if
  end if

  if (percent .neqv. def_percent) then
    if (percent) then
      ret = gptlsetoption (gptlpercent, 1)
    else
      ret = gptlsetoption (gptlpercent, 0)
    end if
  end if

  if (persec .neqv. def_persec) then
    if (persec) then
      ret = gptlsetoption (gptlpersec, 1)
    else
      ret = gptlsetoption (gptlpersec, 0)
    end if
  end if

  if (multiplex .neqv. def_multiplex) then
    if (multiplex) then
      ret = gptlsetoption (gptlmultiplex, 1)
    else
      ret = gptlsetoption (gptlmultiplex, 0)
    end if
  end if

  if (dopr_preamble .neqv. def_dopr_preamble) then
    if (dopr_preamble) then
      ret = gptlsetoption (gptldopr_preamble, 1)
    else
      ret = gptlsetoption (gptldopr_preamble, 0)
    end if
  end if

  if (dopr_threadsort .neqv. def_dopr_threadsort) then
    if (dopr_threadsort) then
      ret = gptlsetoption (gptldopr_threadsort, 1)
    else
      ret = gptlsetoption (gptldopr_threadsort, 0)
    end if
  end if

  if (dopr_multparent .neqv. def_dopr_multparent) then
    if (dopr_multparent) then
      ret = gptlsetoption (gptldopr_multparent, 1)
    else
      ret = gptlsetoption (gptldopr_multparent, 0)
    end if
  end if

  if (dopr_collision .neqv. def_dopr_collision) then
    if (dopr_collision) then
      ret = gptlsetoption (gptldopr_collision, 1)
    else
      ret = gptlsetoption (gptldopr_collision, 0)
    end if
  end if
!
! Character-based variables
!
  if (utr /= def_utr) then
    if (trim(utr) == 'gettimeofday') then
      ret = gptlsetutr (gptlgettimeofday)
    else if (trim(utr) == 'nanotime') then
      ret = gptlsetutr (gptlnanotime)
    else if (trim(utr) == 'rtc') then
      ret = gptlsetutr (gptlrtc)
    else if (trim(utr) == 'mpiwtime') then
      ret = gptlsetutr (gptlmpiwtime)
    else if (trim(utr) == 'clockgettime') then
      ret = gptlsetutr (gptlclockgettime)
    else if (trim(utr) == 'papitime') then
      ret = gptlsetutr (gptlpapitime)
    else
      write(6,*)'gptlprocess_namelist: Underlying timing routine not available: ', utr
    end if
  end if
!
! Print method: use characters for namelist variables to avoid magic numbers
! in namelist
!
  if (print_method /= def_print_method) then
    if (trim(print_method) == 'first_parent') then
      ret = gptlsetoption (gptlprint_method, gptlfirst_parent)
    else if (trim(print_method) == 'last_parent') then
      ret = gptlsetoption (gptlprint_method, gptllast_parent)
    else if (trim(print_method) == 'most_frequent') then
      ret = gptlsetoption (gptlprint_method, gptlmost_frequent)
    else if (trim(print_method) == 'full_tree') then
      ret = gptlsetoption (gptlprint_method, gptlfull_tree)
    else
      write(6,*)'gptlprocess_namelist: print_method not available: ', print_method
    end if
  end if
!
! PAPI-based events
!
  do j=1,maxevents
    if (eventlist(j) /= '                ') then
      ret = gptlevent_name_to_code (trim (eventlist(j)), code)
      if (ret == 0) then
        ret = gptlsetoption (code, 1)
      else
        write(6,*)'gptlprocess_namelist: no code found for event ', eventlist(j)
      end if
    end if
  end do

  close (unit=unitno)
  outret = 0
  return
end subroutine gptlprocess_namelist
